### DISCUSSION TITLE
Legacy FCM API migration

### DISCUSSION THREAD

**Discussion by ostahl** [kreait/firebase-php#911] [General]
Hello, I should fix an issue into an old application using Firebase Notification with your library.

Firebase stop the service and ask for migration https://firebase.google.com/docs/cloud-messaging/migrate-v1?hl=en

Could you tell me which is the oldest version of the library that support this ?
('cause this application run on v4 and php 7.2 ... so it'll be hard to migrate quickly to the last version)

**Comment by jeromegamez** [comment]
Sending single messages should be able with the latest 5.x version. Sending batched/multicast messages is only possible starting with 7.x.

* If you **only** use the `Messaging::send()` method, you will have to upgrade to at least to the latest `5.x` release
* If you also use `Messaging::sendMulticast()` and `Messaging::sendAll()`, you will need to upgrade to at least to the latest `7.x` release
* Version 4 and lower will not work with the new FCM APIs

Full support for the new API could possibly be backported to `6.x`, but that's not trivial, and all but the latest SDK version fall under paid support (=substantial, recurring GitHub Sponsorship)

**Reply by ostahl** [reply]
Thanks a lot for your answer !

**Reply by yodelx** [reply]
Hello,  jeromegames.
Thank you for providing this PHP library.

I have a related question. 
I’ve been using version 5.19 of the library with PHP 7.2, but I’ve noticed an increasing number of error responses with Messaging::sendMulticast(). Since upgrading the PHP version is not an option, I’m considering using Messaging::send(), but I’m encountering performance issues.

Is my understanding correct that the reason Messaging::sendMulticast() and Messaging::sendAll() cannot be used with library versions below 7.0 is due to the deprecation of the batch endpoints following the migration to FCMv1?

Your reply in the following article has been very helpful. I’m considering handling the requests using google/auth for authentication, but I’m concerned that the PHP version might still be a bottleneck.
https://stackoverflow.com/questions/76563436/firebase-http-v1-api-and-no-batch-send-anymore

Thank you,

**Reply by jeromegamez** [reply]
> Is my understanding correct that the reason Messaging::sendMulticast() and Messaging::sendAll() cannot be used with library versions below 7.0 is due to the deprecation of the batch endpoints following the migration to FCMv1?

Yes, you're right. In versions 5.x of the SDK, I migrated the endpoint for sending single messages to the HTTP V1 API. Only in 7.x I migrated the sending of batch messages to the new endpoint as well.

The reason why I didn't do it in 5.x was because at the time, the legacy FCM API was marked as legacy in the docs, but not actually deprecated. Migrating batch sends to also use the new endpoints effectively was not something I was ready to tackle back then, but eventually did fairly recently in 7.x

>  I’m considering using Messaging::send(), but I’m encountering performance issues.

That's exactly the reason why I didn't just change the endpoint for `sendAll()` and `sendMulticast()` - at the time this would have required me to iterate of each message one by one, resulting in the same performance bottlenecks you encounter.

> [...] I’m considering handling the requests using google/auth for authentication, but I’m concerned that the PHP version might still be a bottleneck.

The PHP version will not be a bottleneck as long as your system's cURL has the capability to do HTTP/2 requests. I drop support for PHP outdated versions in major releases of the SDK, but that doesn't mean the features themselves wouldn't work in earlier PHP versions.

Porting the use of the new endpoints to older versions of the SDK is perhaps feasible, but would require a considerable amount of work which I'm not willing to do for free and nobody is willing to pay for it 😅 

tl;dr; The approach in the StackOverflow answer should be working just fine 🤞🏻

**Reply by jeromegamez** [reply]
I need to correct myself: the approach in the StackOverflow response might not be sufficiently performant (it's an old answer 😅). The answer references release 7.5, but the performance issues were only fixed in 7.10 (see https://github.com/kreait/firebase-php/pull/874). You can see performance stats here: https://github.com/kreait/firebase-php/pull/874#issuecomment-2071043551

So, the approach will probably **not** work as it is described there. But it could be a good starting point. If you feel comfortable sifting through source code, you could start looking here https://github.com/kreait/firebase-php/blob/e4b0c0c9be0959cf5f02208fe3815a52f692aebe/src/Firebase/Messaging.php#L85 and perhaps adapt the code in your project.

The SO answer is old and I could rewrite the code and answer text, but I have to be honest, it would take quite some time and my free time is still better used here than there 😅. But perhaps you can add a comment with a link to this discussion so that others might benefit from it. Hopefully 😂

**Reply by yodelx** [reply]
Thank you for your prompt response.

I now have a better understanding of the situation. I will modify the implementation to handle the processing using google/auth and HTTP2 communication. Thank you also for the advice regarding the related ticket.

**Reply by jeromegamez** [reply]
Thank you for your one-time sponsorship 🙏🏻❤️.

If you get stuck somewhere, let me know here in the thread, we'll certainly be able to figure something out!
